INT-3691 Add ZookeeperMetadataStore

mbogoevici opened the pull request:
- create new core interfaces `ListenableMetadataStore` and `MessageStoreListener`;
- create new module `spring-integration-zookeeper`;
- add `ZookeeperMetadataStore` implementation

Commit:
INT-3691 Add ZookeeperMetadataStore

- create new core interfaces `ListenableMetadataStore` and `MessageStoreListener`;
- create new module `spring-integration-zookeeper`;
- add `ZookeeperMetadataStore` implementation

Commit:
Polishing

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/metadata/MessageStoreListenerAdapter.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @author Marius Bogoevici
+ */
+public abstract class MessageStoreListenerAdapter implements MetadataStoreListener {
```
`s/Message/Metadata/` ?

artembilan replied:
+1

mbogoevici replied:
Ouch.

garyrussell commented on spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/Conversions.java:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @author Marius Bogoevici
+ */
+public class Conversions {
```
This should probably go to `-core`; perhaps with a more intuitive name, or maybe just put the methods in `IntegrationUtils` ?

mbogoevici replied:
IntegrationUtils is fine for me.

garyrussell commented on spring-integration-zookeeper/src/test/java/org/springframework/integration/zookeeper/metadata/EventuallyMatcher.java:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @author Eric Bottard
+ */
+public class EventuallyMatcher<U> extends DiagnosingMatcher<U> {
```
Move to `spring-integration-test`

garyrussell commented:
I guess we gave this a thorough review over in the extensions repo so I haven't reviewed closely this time.

Aside from the minor issues above, we need a `zookeeper.adoc` - but we can defer that until we add the other zk support elements if you prefer.

mbogoevici commented:
Thanks for the review - I would prefer to add doc at least after we do namespace support (which is a requirement too).

garyrussell commented:
We don't currently have namespace support for other metadata stores so I don't think this, specifically, needs it.

They are configured as `<bean/>` s.

Commit:
Polishing

- moved classes into core/test where necessary, including conversion methods
- properly renamed `MetadataStoreListenerAdapter`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/metadata/ListenableMetadataStore.java:
```diff
@@ -0,0 +1,25 @@
+package org.springframework.integration.metadata;
```
Oops - needs copyright - will add during merge.

mbogoevici replied:
Argh sorry. Every time I reinstall IntelliJ :(

garyrussell commented:
Merged as 8782d097cf7465efdb69c204ef2f0142b7b1a7cf
